docs/design-proposals/2022-01 Metrics-generator.md

```diff
@@ -0,0 +1,298 @@
+
+#### Metrics-generator ring
+
+The distributor has to find metrics-generator instances present in the cluster. When multiple instances of the metrics-generator are running, the distributor should load balance writes across these instances. Load balancing has to be trace-aware: all spans with the same trace ID should consistently be sent to the same instance.
+
+To achieve this we propose using the [dskit ring](https://github.com/grafana/dskit/tree/main/ring) backed by memberlist. This will be the same mechanism as used by the ingesters. The distributor will shard requests across metrics-generator instances based upon the tokens they own.
+
+_Out-of-scope_: in a later revision we can look into running the metrics-generators with a replication factor of 2 or higher. This is already supported by the ring, but will require extra logic to deduplicate metrics when exporting them (otherwise they are counted multiple times).  
+This is out-of-scope for this design document. Initially the metrics-generator will run with RF=1 only. Note this will result in data loss when an instance crashes.
+
+#### gRPC protocol
+
+Similar to other Tempo components, inter-component requests are sent over gRPC. The existing APIs are defined in [`tempopb/tempo.proto`](../../pkg/tempopb/tempo.proto).
+
+Proposed service definition:
+
+```protobuf
+service MetricsGenerator {
+  rpc PushSpans(PushSpansRequest) returns (PushResponse) {};
+}
+
+// Note: a PushSpansRequest should only contain spans that are relevant to the configured tenants
+// and processors. We can reduce the amount of data sent to the metrics-generator by trimming spans
+// and their metadata in the distributor.
+message PushSpansRequest {
+  // Note: these are full traces. For further optimisation we should consider using a slimmer span
+  // format containing the minimal amount of data.
+  repeated tempopb.trace.v1.ResourceSpans batches = 1;
+}
+
+message PushResponse {
+}
+```
+
+Since the metrics-generator is directly in the write path, an increase in ingress will directly impact the metrics-generator. To reduce the amount of data sent from the distributor to the metrics-generator, the distributor should only send spans that are relevant for the configured metrics processors and tenants. If, for example, a processor only requires a subset of spans the distributor should drop not relevant spans before sending them. This should allow the metrics-generator to scale at a slower rate than the ingesters and saves bandwidth/processing time.
+
+This will require that the distributor is aware of the tenants and processors configured in the metrics-generator. This configuration will thus have to be shared with both components.
```

Comment:
do the processors impact which spans would need  to be sent? my impression is that if we enable metrics-generator for a tenant then all spans will be sent.

Reply:
There is another potential optimisation here: some processors don't need all spans and most processors don't need all tags of a span. So by knowing which processors are enabled and how they are configured, we can trim down the sent data.

2 examples:

- The service graphs processor only uses spans with SpanKind `client` or `server`. All other spans are ignored. If a tenant only uses the service graphs processor, we could drop all other spans. I'd estimate this would easily drop over half of the spans for a specific tenant.

- The span metrics processor uses a limited set of tags as a label in the generated metrics. By default this is the span name, service name, span kind and status. So instead of sending the entire span (which might contain very large tags) we could send a minimal version with just these 4 tags.
  ⚠️ Issue: we want to make the dimensions used by the span metrics processor configurable. That is, a user might want to include tags like `http.target` as well. This setting will have to be shared with the distributor to ensure it doesn't drop this tag accidentally.

Overall this optimisation seems promising, but the implementation is likely to be difficult to maintain and operate. We would have to:

- Share most of the processor config with the distributor, this is necessary because the distributor has to know what to filter out. This will be tricky because some config lives in the `metrics_generator` config block and some in the runtime config. At some point we might also store user config in a bucket, similar to the rules in Cortex/Loki. So that would mean the distributor needs to acces this bucket as well.

- For each processor, we would have to add a filter function. That is, each processor will have to tell us in the distributor whether this span has to be kept or not. This would mean that the implementation of the processor gets spread out over both components making it difficult to maintain.

I think it's a cool idea, but maybe we should keep it simple for now and just say: "the distributor has to know which tenants are included" and keep it at that.